### PR TITLE
Beating JSON performance with Protobuf

### PR DESCRIPTION
**Pull request by brunokrebs**
Outline of the article where I will compare JSON/Protobuf performance.

### CODE REVIEW

File: _posts/2017-01-23-beating-json-performance-with-protobuf.markdown [hunk 1]
```diff
@@ -0,0 +1,337 @@
...
+
+- **Formal format**. Formats are self-describing.
+- **RPC support**. Server RPC interfaces can be declared as part of protocol files.
+- **Structure validation**. Having a predefined and larger, when compared to JSON, set of data types, messages serialized on Protobuf can be automatically validated by the code that is responsible to exchange them.
+
+## How Do We Use Protobuf?
+
+Now that you already know that Protobuf is faster than JSON and you also know its advantages and disadvantages, let's take a look on how do we use this technology. Protobuf has three main components that we have to deal with:
+
+1. **Messages descriptors**. When using Protobuf we have to define our messages structures on `.proto` files.
```
Comment (kukicado) [inline _posts/2017-01-23-beating-json-performance-with-protobuf.markdown]: messages => message

on => in

File: _posts/2017-01-23-beating-json-performance-with-protobuf.markdown [hunk 2]
```diff
@@ -0,0 +1,337 @@
...
+2. **Messages implementations**. Messages definitions are not enough to represent and exchange data on any programming language. We have to generate classes/objects to deal with data on the chosen programming language. Gladly, Google provides code generators for the most common programming languages.
```
Comment (kukicado) [inline _posts/2017-01-23-beating-json-performance-with-protobuf.markdown]: Messages => Message  (Do this for both instances of the word Messages)

on any => in any

on the => in the

Gladly => Luckily

File: _posts/2017-01-23-beating-json-performance-with-protobuf.markdown [hunk 3]
```diff
@@ -0,0 +1,337 @@
...
+3. **Parsing and Serialization**. After defining and creating Protobuf messages, we need to be able to exchange these messages. Google helps us here again, as long as we use one of the supported programming language.
+
+Let's catch a glimpse of each of components.
+
+### Protobuf Message Definition
+
+As already mentioned, messages on Protobuf are describe on `.proto` files. Below you can find an example of the three messages descriptors that I used on my performance tests. I have defined all of them on the same file, which I called `people.proto`.
```
Comment (kukicado) [inline _posts/2017-01-23-beating-json-performance-with-protobuf.markdown]: on => in

messages descriptors => message descriptors

on my => in my

on the same file => in the same file

File: _posts/2017-01-23-beating-json-performance-with-protobuf.markdown [hunk 4]
```diff
@@ -0,0 +1,337 @@
...
+        .addMobile("111111" + i)
+        .addMobile("222222" + i)
+        .addEmail("emailperson" + i + "@somewhere.com")
+        .addEmail("otheremailperson" + i + "@somewhere.com")
+        .addAddress(address1)
+        .addAddress(address2)
+        .build();
+```
+
+These instances alone just represent the messages, so I also needed a way to exchange them. Spring, as always, provides support for Protobuf and there are a few resources out there - like [this one on Spring's blog](https://spring.io/blog/2015/03/22/using-google-protocol-buffers-with-spring-mvc-based-rest-services), and [this one from Baeldung](http://www.baeldung.com/spring-rest-api-with-protocol-buffers) - that helped me on that matter. Just be aware that, as in any Java project, a few dependencies are needed. These are the ones that I had to add to my Maven project:
```
Comment (kukicado) [inline _posts/2017-01-23-beating-json-performance-with-protobuf.markdown]: I would delete "as always"

File: _posts/2017-01-23-beating-json-performance-with-protobuf.markdown [hunk 5]
```diff
@@ -0,0 +1,337 @@
...
+
+## Conclusion
+
+I have to be honest, I was hoping to come across a more favorable scenario for Protobuf. Of course, being able to handle, on a Java to Java communication, 50 thousand instances of `Person` objects in 25ms with Protobuf, while JSON took 150ms, is amazing. But on a JavaScript environment these gains are much lower.
+
+Nevertheless, considering that JSON is native to JavaScript engines, Protobuf still managed to be faster.
+
+Also, one important thing that I noticed is that, even though there are not many resources around about Protobuf, I was still able to use it in different environments without having a hard time. So I guess I will start using this technology with more frequency now.
+
+Are about you? What do you think about the speed of Protobuf? And about starting to use it in your projects? Leave a comment!
```
Comment (kukicado) [inline _posts/2017-01-23-beating-json-performance-with-protobuf.markdown]: Are about you? => How about you?

And about starting... => Are you considering using it in your projects?

File: _posts/2017-01-23-beating-json-performance-with-protobuf.markdown [hunk 6]
```diff
@@ -0,0 +1,337 @@
...
+
+The library used, `protobuf.js`, helped me to compile the `.proto` messages to JavaScript and also to exchange these messages. The first thing that I had to do was to install it as a dependency. For this, I have used [Node.js](https://nodejs.org) and [NPM](https://www.npmjs.com):
+
+```bash
+npm install protobufjs
+```
+
+The command above enabled me to use `pbjs` command line utility (CLI) to generate the code. The following command is how I used this CLI:
+
+```bash
```
Comment (kukicado) [inline _posts/2017-01-23-beating-json-performance-with-protobuf.markdown]: If you have the user install protobufjs globally with the -g flag like `npm install protobuf -g` then they can use `pbjs` by just running `pbjs` and omitting the `.node_modules...` part.

File: _posts/2017-01-23-beating-json-performance-with-protobuf.markdown (hunk 3 above)
Comment (brunokrebs) [inline _posts/2017-01-23-beating-json-performance-with-protobuf.markdown]: Gosh, I have to improve my self in this area. I always make mistakes with prepositions. Thanks

Comment (kukicado) [inline _posts/2017-01-23-beating-json-performance-with-protobuf.markdown]: That's ok. It gets a lot easier as you write more and more 💪

**Pull request merged**